Step2 - 문자열 덧셈 계산기

mins1031 opened the pull request:
안녕하세요
자동차 경주 2단계 과제인 문자열 덧셈 계산기 결과물 입니다
사소한것 하나라도 좋으니 생각할 거리들, 개인적으로 좋다고 생각하시는 기준들을 많이 던져주시면 감사하겠습니다~!

mins1031 committed:
문자열 계산기 구현완료

mins1031 committed:
문자열 계산기 테스트 코드 구현 완료

javajigi commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,84 @@
+        // 숫자가 하나인 경우에도 커스텀 관련 인풋은 인입될 수 있다는 가정하에
+        // 계산 대상 숫자목록 추출작업후 1개 여부 관련 요구사항을 처리했습니다
+        String[] targetNumbers = getNumbers(input);
+        if (hasJustOneNum(targetNumbers)) {
```
hasJustOneNum()에 따라 합을 구하는 로직을 분리해야 할까?
아래 sum 로직과 하나로 구현해도 되지 않을까?

javajigi commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,84 @@
+            return processOneNumber(targetNumbers[0]);
+        }
+
+        return stream(targetNumbers)
+                .mapToInt(this::processOneNumber)
+                .sum();
```
추상화 레벨을 맞추기 위해 이 메서드를 sum()과 같은 메서드로 분리하는 것은 어떨까?

mins1031 replied:
바로 위 피드백 주신 내용과 함께 이해했습니다
말씀주신대로 splitAndSum()의 추상화 레벨을 2단계로 가져가기 위해
getNumbers() 아래 부분을 하나의 메서드로 묶어 처리하는 방향으로 진행하려고 합니다

위 내용대로 해당 두부분을 함께 메서드로 묶는 이유는 상위 메서드의 추상화 레벨을 일정 수준으로 유지하기 위함인데요
제가 생각하지 못한 다른 좋은 이유가 있다면 말씀주시면 감사하겠습니다!

javajigi commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    private int processOneNumber(String targetNumber) {
+        if (isNotDigit(targetNumber)) {
```
Integer.parseInt()에서 숫자가 아닐 경우 NumberFormatException을 throw하는 굳이 숫자인지 여부를 처리해야 할까?

mins1031 replied:
해당 부분 구현시 
try-catch로 NumberFormatException을 잡아서 처리를 할지, 위와 같이 처리를 할지 고민했었는데요
숫자가 아닌 문자가 입력값으로 전달된다는 예외의 상황을 명확하게 처리하는게 좋다고 생각했고 
그러한 맥락에서 try-catch 보다는 명시적으로 처리해주는게 가독성면에서 더 좋다고 판단해 위 처럼 구현했었습니다.

혹시 제가 말씀하신 의도와 다르게 이해했거나 더 효과적인 방향이 있다면 말씀주시면 다시 고민해보겠습니다!

javajigi commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,84 @@
+            throw new RuntimeException("입력값은 숫자여야 합니다");
+        }
+
+        int number = toInt(targetNumber);
+
+        if (isNotPositive(number)) {
+            throw new RuntimeException("입력값은 양수여야 합니다");
+        }
```
toInt() 메서드 이름을 toPositive()와 같은 형태로 바꾸고 음수 여부 처리를  toPositive()에서 확인하는 것은 어떨까?

mins1031 replied:
말씀주신 부분 잘 적용해보았습니다.
적용후에 드는 생각으로는 
1.  processOneNumber() 네이밍을 toInt() 로 바꾸는것 + 기존 toInt()로직 변경. -> int number = Integer.parseInt(input)
2. 말씀주신 음수체크 + 정수변환 

위 두 방식의 어떤 차이가 있는건지 잘 모르겠습니다

첫번째 방향으로 진행하는건 어떻게 생각하시나요??

javajigi commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,84 @@
+        Matcher matcherResult = Pattern.compile(CUSTOM_DELIMITER_REGEX).matcher(input);
+        if (matcherResult.find()) {
+            String delimiter = getCustomDelimiter(matcherResult);
+            String numbers = polishInput(matcherResult);
```
이 메서드의 다른 코드는 추상화 레벨이 0이고, 이 두 부분은 추상화 레벨이 1이다.
굳이 메서드로 분리하기보다 아래와 같이 구현하는 것은 어떻게 생각하나?
어느 부분을 메서드로 분리하는 것이 의미있을지 한번 고민해 보면 어떨까?

```suggestion
            String delimiter = matcherResult.group(1);
            String numbers = matcherResult.group(2);
```

mins1031 replied:
피드백 감사합니다 튜터님.
위 상황에선 말씀하신대로 변수명이 해당 동작에 대한 설명을 충분히 할수 있을것 같아 추상화레벨은 동일하게 가져가는게 좋다는 생각이 듭니다

추가적으로 이런 메서드 분리/클래스 분리에 대한 명확한 기준은 실무에서도 항상 고민되었던 부분인데요.
음.. 제 생각엔 기본적으로는 추상화레벨을 맞추되
1. 도메인적으로 중요한 개념,
2. 2~3 라인이지만 외부 라이브러리를 사용할때
3. 2~3 라인이지만 생소한 기능(ex 복잡한 스트림)or 중요한 핵심 기능일때
위의 케이스에선 추상화 레벨을 올려주는것도 괜찮다는 생각이 드네요

항상 이럴때 마다 저만의 명확한 기준들이 없어서 그때그때 다른 기준을 적용하게 되는것 같아요
어떤날은 A 기준으로 추상화.. 어떤 날은 B 기준으로 추상화..  이렇다보니 제 클린코딩 실력도 늘지 않는것 같다는 생각을 하고 있습니다

뭔가 주절주절 말이 길었던거 같은데..ㅎ..
결론적으로 여쭙고 싶은건 튜터님은 메서드 분리에 어떤 기준을 가지고 계시는지가 궁금합니다

javajigi commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,64 @@
+    private final StringCalculator calculator = new StringCalculator();
+
+    @ParameterizedTest
+    @NullAndEmptySource
```
👍

javajigi reviewed (changes requested):
2단계 미션 진행하느라 수고했어요. 👍 
메서드를 작은 단위로 분리하기 위한 노력의 흔적의 보이네요.
메서드를 극단적으로 분리한 것은 좋은 시도여서 칭찬하고 싶네요.
단, 너무 과도한 분리는 오히려 독이 될 수 있을 것 같아요.
이번 기회에 적정 수준의 메서드 분리는 어느 수준으로 하는 것이 좋을지 생각해 보는 시간을 가지면 어떨까요?

mins1031 committed:
refactor : 피드백 내용기반 개선
- Matcher 그룹기반 로직 추상화 단계 통일
- sumAndSplit 메서드의 추상화레벨 통일을 위해 sum 메서드 분리
- getNumber 메서드 네이밍 개선

javajigi reviewed (approved):
피드백 반영하느라 수고했어요. 👍 
이미 메서드 분리를 잘해 놓은 상태라 크게 개선할 부분이 많지 않았네요.

드디어 본격적인 자동차 경주 게임 구현이네요.
본인의 역량에 따라 어느 부분을 도전할 것인지 고민해 보면 좋겠습니다.
TDD가 생각보다 쉽지 않은 도전이라서요.
기존에 객체 설계, 단위 테스트를 고민해 본 경험이 많지 않다면 3단계는 객체 설계에 집중해도 괜찮아요.
어떤 방식으로 도전했는지 3단계 pr 보낼 때 공유해 주셔도 좋겠습니다.
